Planetary Computer Modular Client

chahibi opened the pull request:
### Packages impacted by this PR

- `@azure/planetarycomputer`

### Issues associated with this PR

N/A

### Describe the problem that is addressed by this PR

This PR introduces a new TypeScript client SDK for the Microsoft Planetary Computer Pro geo-catalog service. The Planetary Computer Pro service provides access to geospatial data through a STAC (SpatioTemporal Asset Catalog) API, enabling users to discover, access, and analyze Earth observation data.

### What are the possible designs available to address the problem? If there are more than one possible design, why was the one in this PR chosen?

N/A

### Are there test cases added in this PR? _(If not, why?)_

In progress

### Provide a list of related PRs _(if any)_

N/A

### Command used to generate this PR:**_(Applicable only to SDK release request PRs)_

Generated from TypeSpec configuration at `specification/orbital/Microsoft.PlanetaryComputer/tspconfig.yaml`

- API Version: 2025-04-30-preview
- SDK Release Type: beta
- CommitSHA: `6caf6d9259e8ee9d5149d88c1857abfa08776543`
- SpecRepo: <https://github.com/Azure/azure-rest-api-specs>

### Checklists

- [ ] Added impacted package name to the issue description
- [ ] Does this PR needs any fixes in the SDK Generator?** _(If so, create an Issue in the [Autorest/typescript](https://github.com/Azure/autorest.typescript) repository and link it here)_
- [ ] Added a changelog (if necessary)

chahibi committed:
Add @azure/planetarycomputer SDK with test recordings

chahibi committed:
Add @azure/planetarycomputer SDK with test recordings

chahibi committed:
Example env vars

chahibi committed:
Example env vars

chahibi committed:
Fixes for browser tests

chahibi committed:
Fixes for browser tests

chahibi committed:
fix(planetarycomputer): update test recordings and fix collection lifecycle race condition

chahibi committed:
fix(planetarycomputer): update test recordings and fix collection lifecycle race condition

chahibi committed:
fix(planetarycomputer): add retry logic for collection lifecycle tests

Added retry with exponential backoff to handle 'collection is being deleted'
race condition when creating collections after deletion. The service's internal
cleanup can take longer than the LRO completion indicates.

Changes:
- test_01 and test_03 now have 30s wait after deletion + 5 retries with backoff
- Increased initial wait from 15s to 30s
- Retries on 'is being deleted' errors with 15s, 30s, 45s, 60s delays

chahibi committed:
fix(planetarycomputer): add retry logic for collection lifecycle tests

Added retry with exponential backoff to handle 'collection is being deleted'
race condition when creating collections after deletion. The service's internal
cleanup can take longer than the LRO completion indicates.

Changes:
- test_01 and test_03 now have 30s wait after deletion + 5 retries with backoff
- Increased initial wait from 15s to 30s
- Retries on 'is being deleted' errors with 15s, 30s, 45s, 60s delays

chahibi committed:
fix(planetarycomputer): fix retry logic for playback mode

In playback mode, the retry count must match the recorded request sequence.
Changed from skipping retries in playback to skipping only the wait times.

This allows the test proxy to replay all recorded retry responses correctly.

chahibi committed:
fix(planetarycomputer): fix retry logic for playback mode

In playback mode, the retry count must match the recorded request sequence.
Changed from skipping retries in playback to skipping only the wait times.

This allows the test proxy to replay all recorded retry responses correctly.

chahibi committed:
restore accidentally deleted ts-http-runtime review files

chahibi committed:
chore(planetarycomputer): regenerate SDK with new TypeSpec commit

- Updated tsp-location.yaml to use commit 52dd71bbf351eb592dca4da3779ad38b51247082
- Renamed listCollections to getCollections in tests
- Updated interval legend test to handle Record<string, any> return type

chahibi committed:
chore(planetarycomputer): regenerate SDK with new TypeSpec commit

- Updated tsp-location.yaml to use commit 52dd71bbf351eb592dca4da3779ad38b51247082
- Renamed listCollections to getCollections in tests
- Updated interval legend test to handle Record<string, any> return type

chahibi committed:
test(planetarycomputer): skip WMTS capabilities tests pending SDK fix

The WMTS capabilities endpoints (_getWmtsCapabilities, _getMosaicsWmtsCapabilities)
have a deserialization bug where XML responses are incorrectly decoded as base64,
resulting in corrupted 12-byte output instead of valid XML.

Skipped tests:
- test_08 in stacItemTiler (STAC Item WMTS Capabilities)
- test_05 in mosaicsTiler (Mosaics WMTS Capabilities)

This is a generated SDK issue that needs to be fixed in the TypeSpec definition
or the emitter.

chahibi committed:
test(planetarycomputer): skip WMTS capabilities tests pending SDK fix

The WMTS capabilities endpoints (_getWmtsCapabilities, _getMosaicsWmtsCapabilities)
have a deserialization bug where XML responses are incorrectly decoded as base64,
resulting in corrupted 12-byte output instead of valid XML.

Skipped tests:
- test_08 in stacItemTiler (STAC Item WMTS Capabilities)
- test_05 in mosaicsTiler (Mosaics WMTS Capabilities)

This is a generated SDK issue that needs to be fixed in the TypeSpec definition
or the emitter.

chahibi committed:
fix(planetarycomputer): fix WMTS XML deserialization bug

- Updated tsp-location.yaml to commit 3397b71c38f969fdf57a433ea669c8fc9f0aa2f7
- Fixed WMTS deserializers to use TextEncoder instead of base64 decoding
  (XML text responses should not be decoded as base64)
- Re-enabled WMTS capability tests that were previously skipped

chahibi committed:
fix(planetarycomputer): fix WMTS XML deserialization bug

- Updated tsp-location.yaml to commit 3397b71c38f969fdf57a433ea669c8fc9f0aa2f7
- Fixed WMTS deserializers to use TextEncoder instead of base64 decoding
  (XML text responses should not be decoded as base64)
- Re-enabled WMTS capability tests that were previously skipped

chahibi committed:
fix(planetarycomputer): remove unused stringToUint8Array import

chahibi committed:
fix(planetarycomputer): remove unused stringToUint8Array import

chahibi committed:
chore(planetarycomputer): update test recordings tag

chahibi committed:
chore(planetarycomputer): update test recordings tag

chahibi committed:
Merge remote-tracking branch 'origin/mpcpro-sdk-gen-pup-js-mod' - resolve pnpm-lock.yaml conflict

chahibi committed:
Merge main

chahibi committed:
Regenerate API

chahibi committed:
Linter issues

chahibi committed:
Unnecessary escape character: \-  no-useless-escape

chahibi committed:
formatting

chahibi committed:
Samples

chahibi committed:
Samples

chahibi committed:
PNPM lock update

chahibi committed:
Merge remote-tracking branch 'origin/main' into mpcpro-sdk-gen-pup-js-mod

chahibi committed:
Merge conflict resolution

chahibi committed:
README url fix

chahibi committed:
Fix links

chahibi committed:
Merge remote-tracking branch 'origin/main' into mpcpro-sdk-gen-pup-js-mod

# Conflicts:
#	.vscode/cspell.json
#	pnpm-lock.yaml

chahibi committed:
Added code owners

chahibi committed:
CI yml file and changelog

chahibi committed:
Add sample configuration and generate samples/v1-beta

chahibi committed:
Fix docs.microsoft.com links: remove /en-us/ locale per CI link checker rules

chahibi committed:
Ignore link

maorleger reviewed (approved):
This looks good for the first beta release! A few comments but I trust you if you want to fix them in a fast-follow PR after merging this

deyaaeldeen commented on sdk/planetarycomputer/planetarycomputer/review/planetarycomputer-api-data-node.api.md:
```diff
@@ -0,0 +1,537 @@
+    colorMapName?: ColorMapNames;
+    dstCrs?: string;
+    expression?: string;
+    format?: TilerImageFormat;
```
I see format is optional in some places and required in others, is that intentional?

deyaaeldeen commented on sdk/planetarycomputer/planetarycomputer/review/planetarycomputer-node.api.md:
```diff
@@ -0,0 +1,2054 @@
+
+// @public
+export interface Point extends Geometry {
+    coordinates: string;
```
Are we sure this is a string and not a `number[]`?

chahibi commented on sdk/planetarycomputer/planetarycomputer/review/planetarycomputer-node.api.md:
```diff
@@ -0,0 +1,2054 @@
+    getCollection: (collectionId: string, options?: StacGetCollectionOptionalParams) => Promise<StacCollection>;
+    getCollectionConfiguration: (collectionId: string, options?: StacGetCollectionConfigurationOptionalParams) => Promise<UserCollectionSettings>;
+    getCollectionQueryables: (collectionId: string, options?: StacGetCollectionQueryablesOptionalParams) => Promise<Record<string, any>>;
+    getCollections: (options?: StacGetCollectionsOptionalParams) => Promise<StacCatalogCollections>;
```
There is no mechanism for pagination in the STAC standard for collection lists

chahibi committed:
Update sdk/planetarycomputer/planetarycomputer/TESTING.md

Co-authored-by: Maor Leger <maorleger@users.noreply.github.com>

chahibi committed:
Update API

chahibi committed:
Merge branch 'mpcpro-sdk-gen-pup-js-mod' of https://github.com/chahibi/azure-sdk-for-js into mpcpro-sdk-gen-pup-js-mod

MaryGao commented on sdk/planetarycomputer/planetarycomputer/review/planetarycomputer-node.api.md:
```diff
@@ -0,0 +1,2054 @@
+// @public
+export interface StacOperations {
+    addMosaic: (collectionId: string, body: StacMosaic, options?: StacAddMosaicOptionalParams) => Promise<StacMosaic>;
+    createCollection: (body: StacCollection, options?: StacCreateCollectionOptionalParams) => PollerLike<OperationState<void>, void>;
```
I think this should be a cross-language issue and we need to update the spec if non-void return is correct. @chahibi does `Operation` represent the real payload for your services? 

Generally the LRO return type is the last-response body payload within its `result` property and see our guideline for [more details](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#obtaining-status-and-results-of-long-running-operations). Therefore if the expected return type is not void we should define the `lroResult` decorator in the [Operation model](https://github.com/Azure/azure-rest-api-specs/blob/e19c31f75d537aa3fb4bd926e02d4968ee83910b/specification/orbital/Microsoft.PlanetaryComputer/inma.tsp#L774-L804) and here is [an example](https://github.com/Azure/azure-rest-api-specs/blob/43ad039a24dc67633aee23176c5213b31aa1bc9f/specification/ai/DocumentIntelligence/models.tsp#L465-L467).


https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#the-status-monitor-resource
<img width="2821" height="1052" alt="image" src="https://github.com/user-attachments/assets/bafe140e-85b5-4957-82d3-bbdc54f97e63" />

chahibi committed:
Workaround for issue 37298

chahibi committed:
Merge branch 'main' into mpcpro-sdk-gen-pup-js-mod

chahibi committed:
Assets

chahibi committed:
Merge branch 'mpcpro-sdk-gen-pup-js-mod' of https://github.com/chahibi/azure-sdk-for-js into mpcpro-sdk-gen-pup-js-mod

chahibi committed:
Address Maor's feedback

chahibi committed:
Address Maor's feedback

chahibi committed:
pnpm lock update

chahibi committed:
Delete

chahibi committed:
Compilation fix

chahibi committed:
npm run format

chahibi commented on sdk/planetarycomputer/planetarycomputer/test/public/utils/byteHelpers.ts:
```diff
@@ -0,0 +1,176 @@
+
+  // Fallback for older Node.js versions
+  // eslint-disable-next-line @typescript-eslint/no-require-imports
+  const zlib = await import("zlib");
```
Removed

chahibi committed:
chore(planetarycomputer): regenerate SDK with TypeSpec commit f6aac155e9

- Updated tsp-location.yaml to use commit f6aac155e9a422cf0251498788e5726c3b615bf3
- Added CONTRIBUTING.md documenting manual modifications needed after regeneration

chahibi committed:
fix(planetarycomputer): apply manual WMTS XML deserialization fix

- Replace base64 decoding with TextEncoder for WMTS XML responses
- Remove unused stringToUint8Array import
- Fixes both getMosaicsWmtsCapabilities and getWmtsCapabilities deserializers

chahibi committed:
chore(planetarycomputer): run format and extract-api

- Format source files with prettier
- Update API review files

chahibi committed:
test(planetarycomputer): fix tests for additionalProperties wrapper

SDK now returns data wrapped in additionalProperties for certain models.
Updated tests to extract data from this wrapper:
- 00_collections.spec.ts: queryables tests
- 02_stacSpecification.spec.ts: queryables test
- 04_stacItemTiler.spec.ts: asset statistics test
- 06_mapLegends.spec.ts: class map and interval legend tests

chahibi committed:
Update typespec version

chahibi committed:
Linter fixes + contributing.md

chahibi committed:
chore: update API review files with type-only imports

chahibi committed:
Invoke "npm run format" inside sdk/planetarycomputer/planetarycomputer to fix formatting

chahibi committed:
Attempt to fix snippet checks

chahibi committed:
Browser test fixes

chahibi committed:
Other attempt to fix code snippet checks

deyaaeldeen commented on sdk/planetarycomputer/planetarycomputer/review/planetarycomputer-node.api.md:
```diff
@@ -0,0 +1,2087 @@
+// @public (undocumented)
+export class PlanetaryComputerProClient {
+    constructor(endpointParam: string, credential: TokenCredential, options?: PlanetaryComputerProClientOptionalParams);
+    readonly data: DataOperations;
```
did you consider a more specific name? `data` feels extremely generic.

deyaaeldeen commented on sdk/planetarycomputer/planetarycomputer/review/planetarycomputer-node.api.md:
```diff
@@ -0,0 +1,2087 @@
+    getTileJson: (collectionId: string, itemId: string, tileMatrixSetId: string, options?: DataGetTileJsonOptionalParams) => Promise<TileJsonMetadata>;
+    getTileMatrixDefinitions: (tileMatrixSetId: string, options?: DataGetTileMatrixDefinitionsOptionalParams) => Promise<TileMatrixSet>;
+    getWmtsCapabilities: (collectionId: string, itemId: string, tileMatrixSetId: string, options?: DataGetWmtsCapabilitiesOptionalParams) => Promise<Uint8Array>;
+    listAvailableAssets: (collectionId: string, itemId: string, options?: DataListAvailableAssetsOptionalParams) => Promise<string[]>;
```
should this be paged? I don't have a sense for how many strings we're talking in the worst case.

deyaaeldeen commented on sdk/planetarycomputer/planetarycomputer/review/planetarycomputer-node.api.md:
```diff
@@ -0,0 +1,2087 @@
+}
+
+// @public
+export interface TilerCoreModelsResponsesPoint {
```
weird name?
